db/migration/migration.go

```diff
@@ -126,25 +120,51 @@ func Run(subCmd string, args []string) error {
 	dbType, dbConnection, migrationsPath, _ := readGooseConfig()
 
 	if err := goose.SetDialect(dbType); err != nil {
-		fmt.Printf("error: failed to set goose dialect: %s\n", err)
-		return err
+		return fmt.Errorf("failed to set goose dialect: %s", err)
 	}
 
 	db, err := sql.Open(dbType, dbConnection)
 
 	if err != nil {
-		fmt.Printf("error: failed to open db: %s\n", err)
-		return err
+		return fmt.Errorf("failed to open db: %s", err)
 	}
 
 	if err = goose.LoadMigrationPlugins(migrationsPath); err != nil {
-		logger.Error("migration: failed to load migration plugins: %s", err)
-		return err
+		return fmt.Errorf("failed to load migration plugins: %s", err)
+	}
+
+	if err := lock(dbType, db); err != nil {
+		return fmt.Errorf("failed to acquire migration lock: %s", err)
+	}
+	defer unlock(dbType, db)
+
+	if err := goose.Run(subCmd, db, migrationsPath, args...); err != nil {
+		return fmt.Errorf("migration failed to run: %s", err)
 	}
 
-	if err = goose.Run(subCmd, db, migrationsPath, args...); err != nil {
-		fmt.Printf("migration: failed to run: %s\n", err)
-		return err
+	return nil
+}
+
+// lock issues a lock on a shared mutex to avoid running migrations on the same
+// database in parallel, currently only supported on mysql.
+func lock(dbType string, db *sql.DB) error {
+	switch dbType {
+	case "mysql":
+		if _, err := db.Exec("LOCK TABLES `goose_db_version` WRITE"); err != nil {
```

Comment:
I'm closing this PR now since the presented solution is wrong in case of Galera cluster and from the ops-team perspective this can be easily handled.